models/events.go

```diff
@@ -1,114 +1,142 @@
-	})
+	// calculate dynamic values
+	cumulativeGasUsed := uint64(0)
+	blockHash, err := e.block.Hash()
+	if err != nil {
+		return nil, err
+	}
 
-	return blocks, nil
+	for i, rcp := range e.receipts {
+		// add transaction hashes to the block
+		e.block.TransactionHashes = append(e.block.TransactionHashes, rcp.TxHash)
+		// calculate cumulative gas used up to that point
+		cumulativeGasUsed += rcp.GasUsed
+		rcp.CumulativeGasUsed = cumulativeGasUsed
+		// set the transaction index
+		rcp.TransactionIndex = uint(i)
+		// set calculate block hash
+		rcp.BlockHash = blockHash
+		// dynamically add missing log fields
+		for _, l := range rcp.Logs {
+			l.TxHash = rcp.TxHash
+			l.BlockNumber = rcp.BlockNumber.Uint64()
+			l.Index = rcp.TransactionIndex
+		}
+	}
+
+	return e, nil
 }
 
-// Transactions finds all the transactions evm events and decodes them into transaction slice,
-// if no transactions is found nil is returned.
-//
-// Return values:
-// []transaction, nil - if transactions are found
-// nil, nil - if no transactions are found
-// nil, err - unexpected error
-func (c *CadenceEvents) Transactions() ([]Transaction, []*StorageReceipt, error) {
-	txs := make([]Transaction, 0)
-	rcps := make([]*StorageReceipt, 0)
+// decodeCadenceEvents accepts Flow Cadence event and decodes it into CadenceEvents
+// collection. It also performs safety checks on the provided event data.
+func decodeCadenceEvents(events flow.BlockEvents) (*CadenceEvents, error) {
+	e := &CadenceEvents{events: events}
 
-	cumulativeGasUsed := uint64(0)
-	var lastReceipt *StorageReceipt
-	for _, e := range c.events.Events {
-		if isTransactionExecutedEvent(e.Value) {
-			tx, receipt, err := decodeTransactionEvent(e.Value)
-			if err != nil {
-				return nil, nil, err
+	// decode and cache block and transactions
+	for _, event := range events.Events {
+		val := event.Value
+
+		if isBlockExecutedEvent(val) {
+			if e.block != nil { // safety check, we can only have 1 or 0 evm blocks per one flow block
+				return nil, fmt.Errorf("EVM block was already set for this Flow block, invalid event data")
 			}
 
-			if lastReceipt != nil && lastReceipt.BlockNumber.Cmp(receipt.BlockNumber) == 0 {
-				cumulativeGasUsed += lastReceipt.GasUsed
-				receipt.CumulativeGasUsed = cumulativeGasUsed
-			} else {
-				cumulativeGasUsed = receipt.GasUsed
-				receipt.CumulativeGasUsed = cumulativeGasUsed
+			block, err := decodeBlockEvent(val)
+			if err != nil {
+				return nil, err
 			}
 
-			lastReceipt = receipt
+			e.block = block
+			continue
+		}
 
-			txs = append(txs, tx)
-			rcps = append(rcps, receipt)
+		if isTransactionExecutedEvent(val) {
+			tx, receipt, err := decodeTransactionEvent(val)
+			if err != nil {
+				return nil, err
+			}
+
+			e.transactions = append(e.transactions, tx)
+			e.receipts = append(e.receipts, receipt)
 		}
 	}
 
-	return txs, rcps, nil
+	// safety check, we can't have an empty block with transactions
+	if e.block == nil && len(e.transactions) > 0 {
```

Comment:
I believe we should add a check to make sure that after all processing, `e.block != nil`, as we should have exactly 1 block for each Flow block, even if without any transactions.